src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,109 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (this.jsonLightInputContext.ReadingResponse
+                && this.jsonLightInputContext.MessageReaderSettings.NullValuesOmitted)
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = GetSelfAndBaseTypeProperties(edmStructuredType);
+                        }
+                        else
+                        {
+                            // Partial subtree. Combine navigation properties and selected properties at the node with distinct.
+                            selectedProperties =
+                                resourceState.SelectedProperties.GetSelectedNavigationProperties(edmStructuredType)
+                                as IEnumerable<IEdmProperty>;
+                            selectedProperties = selectedProperties.Concat(
+                                resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType))
+                                .Distinct();
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+                            if (!currentProperty.Type.IsNullable || currentProperty.PropertyKind == EdmPropertyKind.Navigation)

```

Comment:
>|| currentProperty.PropertyKind == EdmPropertyKind.Navigation [](start = 65, length = 61)

this seems suspect. Doesn't that mean that we wouldn't return a null value for a nav prop that was explicitly selected?  why did we add the nav properties to the selectedProperties list if we just ignore them here?

Reply:
Yes, see comment inside the if block. I believe nav links (@odata.navigationlink annotation) are added by SelectedPropertiesNodes.GetSelectedNavigationProperties method, it looks like a logical method, additionally it has 10 references of src+tests. Filtering out here during the processing of null value restoration should be a reasonable approach?

I will work on ensuring that nested nav property which should be PropertyKind of EdmPropertyKind.Structual can be restored as null.

---
In reply to: [186563420](https://github.com/OData/odata.net/pull/1106#discussion_r186563420) [](ancestors = 186563420)